### ISSUE TITLE
Refactor Puppet code

### ISSUE THREAD

**Issue by novicecpp** [dmwm/CRABServer#7172]
Refactor the puppet code to be easier to maintain.

**Guideline**
- Use current hostgroup, but move machine to crabtaskworker/crabschedd hostgroup
- Maintain backward compatibility, make it easier to migrate and rollback
- `crabdev` will be use to the on `crab-dev-tw{03,04}` machine.
-  Proper format with puppet-lint using [suggestion from reddit user](https://www.reddit.com/r/Puppet/comments/opcgwb/comment/h6jvdl0/?utm_source=reddit&utm_medium=web2x&context=3)

Tasklist
- [x] Remove setup VM scripts from dmwm/deployment)
- [ ] Refactor hiera data according host group instead of put it on every fqdn file.
  - [x] `sendmail::root_email` to hostgroup's hiera level.
- [x] add ssh pubkey to crab3 for jenkins deploy new taskworker/publisher image.
- [ ] review all cronjob
  - [x] cleanTW.sh should we run every week instead of every month?
- [x] [Make sure that membership in docker unix group in crab-*tw* VM's is managed via e-group](https://github.com/dmwm/CRABServer/issues/7150)
- [x] Install Schedd token to TW machine

**Comment by novicecpp** [comment]
- [ ] tidy up hostgroup https://gitlab.cern.ch/ai/it-puppet-hostgroup-vocmsglidein/-/merge_requests/59
@mapellidario Please review.

**Comment by novicecpp** [comment]
https://gitlab.cern.ch/ai/it-puppet-hostgroup-vocmsglidein/-/merge_requests/65
I already merged this. It has been tested when I deployed TaskWorker/Publisher in `crab-dev-tw02`. It only run first time when machine start (by checking `/data` directory).

What those scripts do are:
- add local user (already add in container).
- request service certificates (puppet handle this).
- make all sudoer to not supply password when run `sudo` command (we got `*** SECURITY information for crab-dev-tw04.cern.ch ***` email because of this).
- install some package (already do in container).

**Comment by novicecpp** [comment]
- [ ] MR: [Move TaskWorker Machine to crabtaskworker sub-hostgroup](https://gitlab.cern.ch/ai/it-puppet-hostgroup-vocmsglidein/-/merge_requests/66)

**Comment by novicecpp** [comment]
- [ ] MR: [use subhostgroup instead of profile](https://gitlab.cern.ch/ai/it-puppet-hostgroup-vocmsglidein/-/merge_requests/67) will be merge after [MR#66](https://gitlab.cern.ch/ai/it-puppet-hostgroup-vocmsglidein/-/merge_requests/66)

**Comment by novicecpp** [comment]
[@belforte](https://github.com/belforte) Do you know why we change how we deploy TaskWorker/Publisher configuration in Puppet? From pulling configuration file from [crab3/CRAB3SeverConfig](https://gitlab.cern.ch/crab3/CRAB3ServerConfig/-/tree/master/TaskWorker/config) To Puppet [Hiera data](https://gitlab.cern.ch/ai/it-puppet-hostgroup-vocmsglidein/-/blob/c26a821ec5e97e10183c210c35a1315b3e049020/data/fqdns/crab-prod-tw01.cern.ch.yaml#L7-33)?

I think pulling configuration from CRAB3SeverConfig repo is better. We do not need to maintain puppet code when configuration change. It is the configuration for our app, which we know every single line on how it uses inside our code.

**Comment by belforte** [comment]
I think the point here is that there are two origins for changes in the TaskWorker/Publisher config files:
1. changes due to the code : which as you say would suggest to keep file in GH
2. changes needed to customise the configuration for each particular deployment, which suggest to have an ERB file in puppet

so the current situation is due to having preferred to make 2. easy.
 All in all changes to the part of this file which is not templated in erb/hiera are very rare, so the work which you indicate is annoying but not really delaying progress.
I agree that there could be other ways. I am open to ideas but atm do not see what exactly you propose.

**Comment by novicecpp** [comment]
I want to to change back to "1.". We can pull config by instance name instead of environment. It same thing we are doing in each hiera fqdn right now without have extra puppet code.

But well, I think we can do it later if we decide to revise/refactor the configuration file somehow.

**Comment by belforte** [comment]
do you mean to have in GH one config. file for every TW instance ? And possibly create them automatically from a common templace and a "localization" script ? Yes, that would work.

**Comment by novicecpp** [comment]
>do you mean to have in GH one config. file for every TW instance ?

Yes.

> And possibly create them automatically from a common templace and a "localization" script

No, just copy/paste/replace. If we really want to do some common template, we should change at crab code relate to configuration instead.

**Comment by novicecpp** [comment]
Note about how do I test change on TW machine:

- Reset `crabdev` to `crabqa`.
- Merge change to `crabdev`.
- Run `puppet agent -tv` on `crab-dev-tw03`.
- Run `ai-kill crab-dev-tw04.cern.ch`.
  - I did not check if all puppet change are apply, only look error report in foreman.

**Comment by novicecpp** [comment]
- [ ] changes solving some tasks in tasklist:  https://gitlab.cern.ch/ai/it-puppet-hostgroup-vocmsglidein/-/merge_requests/68

**Comment by novicecpp** [comment]
I have reconciled qa to master branch in new [master_reconcile](https://gitlab.cern.ch/ai/it-puppet-hostgroup-vocmsglidein/-/tree/master_reconcile) branch, suggested by Nikos, to make `qa` branch available to use again. He also confirmed that we did not need to put machine into `qa` branch/puppet-environment to let machine able to connect ITB pool.
- [diff master_reconcile to master](https://gitlab.cern.ch/ai/it-puppet-hostgroup-vocmsglidein/-/compare/d7c30d17572b942b3bc760ac5a8f6fbfc1a57c26...4f7eeaa275470848ca7aa71ffc61e64dccda53e2?from_project_id=4758)
- [diff master_reconcile to qa](https://gitlab.cern.ch/ai/it-puppet-hostgroup-vocmsglidein/-/compare/ef227cc6bbdaebb70b3b9acc131ead62c19f738a...4f7eeaa275470848ca7aa71ffc61e64dccda53e2?from_project_id=4758)

Old qa branch is here: [qa_before_reconcile](https://gitlab.cern.ch/ai/it-puppet-hostgroup-vocmsglidein/tree/qa_before_reconcile/)

------------

Step for reconcile branch:
- Create new branch `master_reconcile` start at origin/master.
- Checkout `master_reconcile`, merge `qa` branch.
  - Main point to merge `qa` into `master_reconcile` is to check if there any update from qa that do not apply to master yet.
  - Check each conflicted files if it should apply `qa` change to `master`. Usually not except some code relate to vocms068/069 (e.g. data/fqdn files, manifests/node.pp).  
    - Luckily, we have less changes than I though.
- Check diff between `master_reconcile` and `master` if there introduce new change in production node (which is none). 

Next step would be 
- Merge `master_reconcile` back to `qa`.
- Run `puppet agent -tv`  and check how vocms068/069 behave.
  - If something go wrong, check diff `master_reconcile` to `qa` in link above.

**Comment by novicecpp** [comment]
Need to put vocms059 to `qa` branch to check if anything break in production machine before merge to `master`.

**Comment by novicecpp** [comment]
Update:
- On production
  - reconciled master branch.
  - [!68](https://gitlab.cern.ch/ai/it-puppet-hostgroup-vocmsglidein/-/merge_requests/68), to solve some of task list (see in MR description).
  - [!69](https://gitlab.cern.ch/ai/it-puppet-hostgroup-vocmsglidein/-/merge_requests/69), new schedd machine,
  - [!70](https://gitlab.cern.ch/ai/it-puppet-hostgroup-vocmsglidein/-/merge_requests/70), move to new sub-hostgroup.
- Last MR, refactor schedd [!71](https://gitlab.cern.ch/ai/it-puppet-hostgroup-vocmsglidein/-/merge_requests/71), reviewed by Dario, is merged into `qa`.

**Comment by novicecpp** [comment]
I merged `qa` to `production` yesterday, see https://cms-logbook.cern.ch/elog/Analysis+Operations/3780.

2 tasks I did not do in this refactor
- `Refactor hiera data according host group instead of put it on every fqdn file`. 
  - Well, I do not know what decision to move from fqdn to hostgroup. So, leave it like this may be not bad.
- `review all cronjob` 
  - I think it will take a lot of time and effort to understand why and how it affect taskworker/schedd, especially in schedd machine to make sure it will not make job/schedd fail.
  
And 1 task left before I can close this issues:
- [ ] Update puppet docs.
